cql3/statements/select_statement.cc

```diff
@@ -1415,9 +1415,22 @@ indexed_table_select_statement::find_index_partition_ranges(query_processor& qp,
         ));
     }
     size_t max_vector_size = is_paged ? MAX_PARTITION_RANGES : std::numeric_limits<size_t>::max();
+    std::unique_ptr<query_options> paging_adjusted_options = [&] () {
+        if (is_paged && _schema->clustering_key_size() == 0 && static_cast<size_t>(options.get_page_size()) > max_vector_size) {
+            // Result is guaranteed to contain one partition key per row.
+            // Limit the page size a priori.
+            auto internal_options = std::make_unique<cql3::query_options>(cql3::query_options(options));
+            internal_options.reset(new cql3::query_options(std::move(internal_options), options.get_paging_state(), max_vector_size));
+            return internal_options;
+        }
+        return std::unique_ptr<cql3::query_options>(nullptr);
+    }();
+    return do_with(std::move(paging_adjusted_options), [&, max_vector_size](auto& paging_adjusted_options) {
+    // FIXME: Indent.
```

Comment:
Please follow up (no need to do so in this series) with a coroutinization of this function, we should not be adding do_with:s.